File: compiler.py
# Python code to convert assembly to machine code

from collections import defaultdict 
import struct
from sys import argv

# Warning: this register might be arbitrarily overwritten due to intermediate instructions
TEMPORARY_REGISTER = "x30"

register_equivalence = {
    "zero": "x0",
    "ra": "x1",
    "sp": "x2",
    "gp": "x3",
    "tp": "x4",
    "t0": "x5",
    "t1": "x6",
    "t2": "x7",
    "s0": "x8",
    "s1": "x9",
    "a0": "x10",
    "a1": "x11",
    "a2": "x12",
    "a3": "x13",
    "a4": "x14",
    "a5": "x15",
    "a6": "x16",
    "a7": "x17",
    "s2": "x18",
    "s3": "x19",
    "s4": "x20",
    "s5": "x21",
    "s6": "x22",
    "s7": "x23",
    "s8": "x24",
    "s9": "x25",
    "s10": "x26",
    "s11": "x27",
    "t3": "x28",
    "t4": "x29",
    "t5": "x30",
    "t6": "x31",
    "ft0": "f0",
    "ft1": "f1",
    "ft2": "f2",
    "ft3": "f3",
    "ft4": "f4",
    "ft5": "f5",
    "ft6": "f6",
    "ft7": "f7",
    "fs0": "f8",
    "fs1": "f9",
    "fa0": "f10",
    "fa1": "f11",
    "fa2": "f12",
    "fa3": "f13",
    "fa4": "f14",
    "fa5": "f15",
    "fa6": "f16",
    "fa7": "f17",
    "fs2": "f18",
    "fs3": "f19",
    "fs4": "f20",
    "fs5": "f21",
    "fs6": "f22",
    "fs7": "f23",
    "fs8": "f24",
    "fs9": "f25",
    "fs10": "f26",
    "fs11": "f27",
    "ft8": "f28",
    "ft9": "f29",
    "ft10": "f30",
    "ft11": "f31",
}


funct3_map = {
    "JALR": "000",
    "BEQ": "000",
    "BNE": "001",
    "BLT": "100",
    "BGE": "101",
    "BLTU": "110",
    "BGEU": "111",
    "LB": "000",
    "LH": "001",
    "LW": "010",
    "LBU" : "100",
    "LHU" : "101",
    "SB" : "000",
    "SH" : "001",
    "SW" : "010",
    "ADDI" : "000",
    "SLTI" : "010",
    "SLTIU" : "011",
    "XORI" : "100",
    "ORI" : "110",
    "ANDI" : "111",
    "SLLI" : "001",
    "SRLI" : "101",
    "SRAI" : "101",
    "ADD" : "000",
    "SUB" : "000",
    "MUL" : "000",
    "SLL" : "001",
    "SLT" : "010",
    "SLTU" : "011",
    "XOR" : "100",
    "SRL" : "101",
    "SRA" : "101",
    "OR" : "110",
    "AND" : "111",
    "FADD.S" : "000",
    "FSUB.S" : "000",
    "FMUL.S" : "000",
    "FEQ.S" : "010",
    "FLT.S" : "001",
    "FLE.S" : "000",
    "FMV.X.W" : "000",
    "FMV.W.X" : "000",
    "FSGNJ.S" : "000",
    "FSGNJN.S" : "001",
    "FSGNJX.S" : "010",
    "FCVT.S.W" : "000",
    "FLW" : "010",
    "FSW" : "010",
}

funct7_map = defaultdict(lambda: "0000000")
funct7_map.update({
    "SRAI": "0100000",
    "SUB" : "0100000",
    "MUL" : "0000001",
    "SRA" : "0100000",
    "FADD.S" : "0000000",
    "FSUB.S" : "0000100",
    "FMUL.S" : "0001000",
    "FEQ.S" : "1010000",
    "FLT.S" : "1010000",
    "FLE.S" : "1010000",
    "FMV.X.W" : "1110000",
    "FMV.W.X" : "1111000",
    "FSGNJ.S" : "0010000",
    "FSGNJN.S" : "0010000",
    "FSGNJX.S" : "0010000",
    "FCVT.S.W" : "1101000",
})

def int2signedbin(number: int, digits: int) -> int:
    assert -2 ** digits < number < 2 ** digits 
    if number >= 0:
        return number
    else:
        xor_mask = int("1"*digits, 2)
        two_complement = (abs(number) ^ xor_mask) + 1
        return two_complement

def float2signedbin(value: float) -> int:
    packed_value = struct.pack('>f', value)  # big-endian, 32-bit float
    # return ''.join(f'{byte:08b}' for byte in packed_value) # String
    return struct.unpack('>i', packed_value)[0]

def divide_lui_addi(value: int) -> tuple[int, int]:
    # Parte of the "load immediate" chain
    # Divides the 32-bit integer into two immediates that go to lui and addi instruction
    lui_int = value >> 12

    addi_int = value & 0xFFF # Take 12 bits
    # ADDI is sign-extended, so a small compensation might be needed
    if (value & 0x800) != 0: # Takes the 12-th bit
        lui_int += 1 # Annulates the sign-extension of ADDI
        addi_int -= 0x1000 # The immediate will be negative
    
    return lui_int, addi_int

print(len(argv))

if len(argv) >= 2:
    file_r_name = argv[1]
else:
    file_r_name = 'input.txt'

if len(argv) >= 3:
    file_w_name = argv[2]
else:
    file_w_name = 'output.txt'

file_r = open(file_r_name, 'r')
file_w = open(file_w_name, 'w')

Lines = file_r.readlines()
gotos = {}
counter_goto = 0
for line in Lines:
    line = line.strip().rstrip("\n").upper()
    words = line.split(' ')
    opcode = words[0]
    if opcode == "":
        continue
    elif opcode[-1] != ":":
        counter_goto += 4
    else:
        gotos[opcode.rstrip(":")] = counter_goto



print("Gotos:", gotos)

counter = 0
def process_line(line):

    global counter

    line = line.strip().rstrip("\n").upper()
    words = line.split(' ')
    words = [w.rstrip().rstrip(',') for w in words]
    opcode = words[0]
    opcode = opcode.upper()
    bytecode = ""

    if opcode == "":
        return
    if opcode == "CSRRW": #XXX: Ignore this instruction for now
        opcode = "NOP"
    if opcode[-1] == ":":
        assert(gotos[opcode.rstrip(":")] == counter)
        return
    
    #Pseudoinstructions
    if opcode == "MOV" or opcode == "MV":
        try:
            int(words[2])
            #TODO: Make it work with numbers bigger than 2**12
            words = ["ADDI", words[1], "x0", words[2]]
        except ValueError:
            words = ["ADDI", words[1], words[2], "0"]
    elif opcode == "NOP":
        words = ["ADDI", "x0", "x0", "0"]
    elif opcode == "FMOV" or opcode == "FMV" or opcode == "FMOV.S" or opcode == "FMV.S":
        try:
            #XXX: Warning: this uses the register X30
            float_imm = float(words[2])
            float_bin = float2signedbin(float_imm)
            lui_int, addi_int = divide_lui_addi(float_bin)
            process_line(f"LUI {TEMPORARY_REGISTER} {lui_int}")
            process_line(f"ADDI {TEMPORARY_REGISTER} {TEMPORARY_REGISTER} {addi_int}")
            words = ["FMV.W.X", words[1], TEMPORARY_REGISTER]
        except ValueError: # Move between FP registers
            words = ["FSGNJ.S", words[1], words[2], words[2]]

    # Reload opcode
    opcode = words[0]

    if '(' in words[2]:
        words.append((words[2].split('('))[0])
        words[2] = (words[2].split('('))[1][:-1]

    # Adjust register names
    for i, word in enumerate(words):
        if word.lower() in register_equivalence:
            words[i] = register_equivalence[word.lower()]

    match opcode:
        case "LUI":
            rd = int(words[1][1:])
            imm = int(words[2])
            imm = int2signedbin(imm, 20)
            opcode_number = int("0110111", 2)
            bytecode = '{:020b}{:05b}{:07b}'.format(imm, rd, opcode_number)
        case "AUIPC":
            rd = int(words[1][1:])
            imm = int(words[2])
            imm = int2signedbin(imm, 20)
            opcode_number = int("0010111", 2)
            bytecode = '{:020b}{:05b}{:07b}'.format(imm, rd, opcode_number)
        case "JAL":
            rd = int(words[1][1:])
            imm = int(words[2])
            imm = int2signedbin(imm, 30)
            opcode_number = int("1101111", 2)
            binary_imm = '{:030b}'.format(imm)[::-1]
            formated_binary_imm = binary_imm[20] + binary_imm[10:0:-1] + binary_imm[11] + binary_imm[19:11:-1]
            bytecode = '{}{:05b}{:07b}'.format(formated_binary_imm, rd, opcode_number)
        case "JALR":
            rd = int(words[1][1:])
            rs1 = int(words[2][1:])
            imm = int(words[3])
            imm = int2signedbin(imm, 12)
            opcode_number = int("1100111", 2)
            bytecode = '{:012b}{:05b}{}{:05b}{:07b}'.format(imm, rs1, funct3_map[opcode],rd, opcode_number)
        case "BEQ"|"BNE"|"BLT"|"BGE"|"BLTU"|"BGEU":
            rs1 = int(words[1][1:])
            rs2 = int(words[2][1:])
            imm = int(words[3])
            imm = int2signedbin(imm, 30)
            opcode_number = int("1100011", 2)
            binary_imm = '{:030b}'.format(imm)[::-1]
            formated_binary_imm1 = binary_imm[12] + binary_imm[10:4:-1]
            formated_binary_imm2 = binary_imm[4:0:-1] + binary_imm[11]
            bytecode = '{}{:05b}{:05b}{}{}{:07b}'.format(formated_binary_imm1, rs2, rs1, funct3_map[opcode], formated_binary_imm2, opcode_number)
        case "LB"|"LH"|"LW"|"LBU"|"LHU":
            rd = int(words[1][1:])
            rs1 = int(words[2][1:])
            imm = int(words[3])
            imm = int2signedbin(imm, 12)
            opcode_number = int("0000011", 2)
            bytecode = '{:012b}{:05b}{}{:05b}{:07b}'.format(imm, rs1, funct3_map[opcode], rd, opcode_number)
        case "SB"|"SH"|"SW":
            rs2 = int(words[1][1:])
            rs1 = int(words[2][1:])
            imm = int(words[3])
            imm = int2signedbin(imm, 12)
            opcode_number = int("0100011", 2)
            binary_imm = '{:012b}'.format(imm)[::-1]
            formated_binary_imm1 =  binary_imm[11:4:-1]
            formated_binary_imm2 = binary_imm[4::-1]
            bytecode = '{}{:05b}{:05b}{}{}{:07b}'.format(formated_binary_imm1, rs2, rs1, funct3_map[opcode], formated_binary_imm2, opcode_number)
        case "ADDI"|"SLTI"|"SLTIU"|"XORI"|"ORI"|"ANDI":
            rd = int(words[1][1:])
            rs1 = int(words[2][1:])
            imm = int(words[3])
            imm = int2signedbin(imm, 12)
            opcode_number = int("0010011", 2)
            bytecode = '{:012b}{:05b}{}{:05b}{:07b}'.format(imm, rs1, funct3_map[opcode], rd, opcode_number)
        case "SLLI"|"SRLI"|"SRAI":
            rd = int(words[1][1:])
            rs1 = int(words[2][1:])
            imm = int(words[3])
            assert (imm >= 0)
            imm = int2signedbin(imm, 5)
            opcode_number = int("0010011", 2)
            bytecode = '{}{:05b}{:05b}{}{:05b}{:07b}'.format(funct7_map[opcode], imm, rs1, funct3_map[opcode], rd, opcode_number)
        case "ADD"|"SUB"|"MUL"|"SLL"|"SLT"|"SLTU"|"XOR"|"SRL"|"SRA"|"OR"|"AND":
            rd = int(words[1][1:])
            rs1 = int(words[2][1:])
            rs2 = int(words[3][1:])
            opcode_number = int("0110011", 2)
            bytecode = '{}{:05b}{:05b}{}{:05b}{:07b}'.format(funct7_map[opcode], rs2, rs1, funct3_map[opcode], rd, opcode_number)
        case "FADD.S"|"FSUB.S"|"FMUL.S"|"FEQ.S"|"FLT.S"|"FLE.S"|"FMV.X.W"|"FMV.W.X"|"FSGNJ.S"|"FSGNJN.S"|"FSGNJX.S"|"FCVT.S.W":
            rd = int(words[1][1:])
            rs1 = int(words[2][1:])
            rs2 = int(words[3][1:]) if len(words) > 3 else 0
            opcode_number = int("1010011", 2)
            bytecode = '{}{:05b}{:05b}{}{:05b}{:07b}'.format(funct7_map[opcode], rs2, rs1, funct3_map[opcode], rd, opcode_number)
        case "FLW":
            rd = int(words[1][1:])
            rs1 = int(words[2][1:])
            imm = int(words[3])
            imm = int2signedbin(imm, 12)
            opcode_number = int("0000111", 2)
            bytecode = '{:012b}{:05b}{}{:05b}{:07b}'.format(imm, rs1, funct3_map[opcode], rd, opcode_number)
        case "FSW":
            rs2 = int(words[1][1:])
            rs1 = int(words[2][1:])
            imm = int(words[3])
            imm = int2signedbin(imm, 12)
            opcode_number = int("0100111", 2)
            binary_imm = '{:012b}'.format(imm)[::-1]
            formated_binary_imm1 =  binary_imm[11:4:-1]
            formated_binary_imm2 = binary_imm[4::-1]
            bytecode = '{}{:05b}{:05b}{}{}{:07b}'.format(formated_binary_imm1, rs2, rs1, funct3_map[opcode], formated_binary_imm2, opcode_number)
        case _:
            raise ValueError("Opcode not found: ", opcode)

    print(words)
    print(bytecode)
    assert (len(bytecode) == 32)
    file_w.write(bytecode + "\n")
    counter += 4

file_r = open(file_r_name, 'r')
Lines = file_r.readlines()
for line in Lines:
    process_line(line)
file_r.close()



print(f"Finished decoding from {file_r_name} to {file_w_name}")
input()

File: compiler.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <ctype.h>
#include <assert.h>

#define FUNCT7_DEFAULT 0
#define FUNCT3_MAP_SIZE 33
#define FUNCT7_MAP_SIZE 3
#define MAX_LINE_LENGTH 128


// Define the endinanness of the output:
// Possibilities:
// LITTLE, MIXED, BIG  
// MIXED corresponds to LITTLE in memory addresses and BIG inside the bytes
//ENDIANNESS = MIXED;
enum endianness_type{LITTLE = 0, MIXED = 1, BIG = 2};
enum endianness_type ENDIANNESS = MIXED;




// Function to get the funct3 code
const int get_funct3(const char* opcode) {
    static const char* funct3_map[FUNCT3_MAP_SIZE][2] = {
        {"BEQ", "000"}, {"BNE", "001"}, {"BLT", "100"}, {"BGE", "101"}, {"BLTU", "110"},
        {"BGEU", "111"}, {"LB", "000"}, {"LH", "001"}, {"LW", "010"}, {"LBU", "100"},
        {"LHU", "101"}, {"SB", "000"}, {"SH", "001"}, {"SW", "010"}, {"ADDI", "000"},
        {"SLTI", "010"}, {"SLTIU", "011"}, {"XORI", "100"}, {"ORI", "110"}, {"ANDI", "111"},
        {"SLLI", "001"}, {"SRLI", "101"}, {"SRAI", "101"}, {"ADD", "000"}, {"SUB", "000"},
        {"SLL", "001"}, {"SLT", "010"}, {"SLTU", "011"}, {"XOR", "100"}, {"SRL", "101"},
        {"SRA", "101"}, {"OR", "110"}, {"AND", "111"}
    };

    for (int i = 0; i < FUNCT3_MAP_SIZE; ++i) {
        if (strcmp(funct3_map[i][0], opcode) == 0) {
            return atoi(funct3_map[i][1]);
        }
    }
    return 0;
}

// Function to get the funct7 code
const int get_funct7(const char* opcode) {
    static const char* funct7_map[FUNCT7_MAP_SIZE][2] = {
        {"SRAI", "0100000"}, {"SUB", "0100000"}, {"SRA", "0100000"}
    };

    for (int i = 0; i < FUNCT7_MAP_SIZE; ++i) {
        if (strcmp(funct7_map[i][0], opcode) == 0) {
            return atoi(funct7_map[i][1]);
        }
    }
    return FUNCT7_DEFAULT;
}

int int2signedbin(int number, int digits) {
    if (number >= 0) {
        return number;
    } else {
        int xor_mask = (1 << (digits)) - 1; // mask with ones for bit inversion
        int two_complement = (labs(number) ^ xor_mask) + 1; // 2's complements
        return two_complement;
    }
}

void byte2bits(char * bits, const int byte_in)
{
    int i;
    for (i = 0; i <= 7; i++) {
        if (ENDIANNESS == MIXED)
            bits[i] = ((byte_in >> (7-i)) & 1) + '0';
    }
}

int main() {
    
    FILE *file_r, *file_w;
    char line[MAX_LINE_LENGTH];
    int counter = 0;
    int gotos_size = 0;
    char gotos[MAX_LINE_LENGTH][MAX_LINE_LENGTH];

    // Open input and output files
    file_r = fopen("input.txt", "r");
    file_w = fopen("output.txt", "w");
    if (file_r == NULL || file_w == NULL) {
        printf("Error opening files.\n");
        return 1;
    }

    // Read lines from input file
    while (!feof(file_r)) {
        if (fgets(line, MAX_LINE_LENGTH, file_r) == NULL)
            continue;
        if (strlen(line) <= 2) // Remove buggy lines
            continue;
        
        // Remove leading/trailing whitespace and convert to uppercase
        char *p = line;
        while (*p && isspace(*p)) p++;
        size_t len = strlen(p);
        while (len > 0 && isspace(p[len - 1])) len--;
        p[len] = '\0';
        for (char *q = p; *q; q++) {
            *q = toupper(*q);
        }

        // Split the line into words
        char *words[MAX_LINE_LENGTH];
        char *token = strtok(p, " ");
        int word_count = 0;
        while (token != NULL) {
            words[word_count++] = token;
            token = strtok(NULL, " ");
        }

        // Process opcode and update gotos
        char *opcode = words[0];
        if (strcmp(opcode, "") == 0) {
            continue; // Skip empty lines
        } else if (opcode[strlen(opcode) - 1] != ':') {
            counter++;
        } else {
            opcode[strlen(opcode) - 1] = '\0'; // Remove colon
            strncpy(gotos[gotos_size], opcode, MAX_LINE_LENGTH);
            gotos_size++;
        }
    }

    // Print gotos
    printf("Gotos:");
    for (int i = 0; i < gotos_size; i++) {
        printf(" %s", gotos[i]);
    }
    printf("\n");

    counter = 0;
    int bytecounter = 0;

    // Open input file
    file_r = fopen("input.txt", "r");
    if (file_r == NULL) {
        printf("Error opening input file.\n");
        return 1;
    }

    // Read lines from input file
    while (!feof(file_r)) {
        if (fgets(line, MAX_LINE_LENGTH, file_r) == NULL)
            continue;
        if (strlen(line) <= 2) // Remove buggy lines
            continue;
        
        // Remove leading/trailing whitespace and convert to uppercase
        char *p = line;
        while (*p && isspace(*p)) p++;
        size_t len = strlen(p);
        while (len > 0 && isspace(p[len - 1])) len--;
        p[len] = '\0';
        for (char *q = p; *q; q++) {
            *q = toupper(*q);
        }

        // Split the line into words
        char *words[MAX_LINE_LENGTH];
        char *token = strtok(p, " ");
        int word_count = 0;
        while (token != NULL) {
            words[word_count++] = token;
            token = strtok(NULL, " ");
        }

        // Convert assembly format
        char *parenthesis = strchr(words[2], '(');
        if (parenthesis != NULL) {
            words[3] = words[2];
            words[2] = &parenthesis[1];
            *parenthesis = '\0'; // Remove parenthesis
            words[2][strlen(words[2]) - 1] = '\0'; // Remove parenthesis
        }

        // Get the opcode
        char *opcode = words[0];
        if (strcmp(opcode, "") == 0) {
            continue; // Skip empty lines
        } else if (opcode[strlen(opcode) - 1] != ':') {
            counter++;
        } else {
            continue; // Skip gotos
        }

        // Process pseudoinstructions
        if (strcmp(opcode, "MOV") == 0 || strcmp(opcode, "MV") == 0) {
            opcode = "ADDI";
            if ((words[2][0] >= '0') && (words[2][0] <= '9')) {
                words[0] = "ADDI";
                strcpy(words[3], words[2]);
                words[2] = "x0";
                
            } else {
                words[0] = "ADDI";
                words[3] = "0";
            }
            word_count++;
        } else if (strcmp(opcode, "NOP") == 0) {
            opcode = "ADDI";
            words[0] = "ADDI";
            words[1] = "x0";
            words[2] = "x0";
            words[3] = "0";
        }

        int bytecode;

        if (strcmp(opcode, "") == 0) {
            continue;
        } else if (opcode[strlen(opcode) - 1] == ':') {
            // assert(gotos[opcode]); // Verify if goto corresponds to registered value
        } else {
            if (strcmp(opcode, "LUI") == 0) {
                int rd = atoi(&words[1][1]);
                int imm = atoi(words[2]);
                imm = int2signedbin(imm, 20);
                int opcode_number = 0b0110111;
                bytecode = ((imm & 0xFFFFF) << 12) | ((rd & 0x1F) << 7) | (opcode_number & 0x7F);
            } else if (strcmp(opcode, "AUIPC") == 0) {
                int rd = atoi(&words[1][1]);
                int imm = atoi(words[2]);
                if (words[2][0] == '-') imm *= -1;
                imm = int2signedbin(imm, 20);
                int opcode_number = 0b0010111;
                bytecode = ((imm & 0xFFFFF) << 12) | ((rd & 0x1F) << 7) | (opcode_number & 0x7F) ;
            } else if (strcmp(opcode, "JAL") == 0) {
                int rd = atoi(&words[1][1]);
                int imm = atoi(words[2]);
                imm = int2signedbin(imm, 30);
                int opcode_number = 0b1101111;
                // Generate bytecode
                int binary_imm = (((imm >> 1) & 0x3FF) << 21) | (((imm >> 11) & 0x1) << 20) | (((imm >> 12) & 0xFF) << 12) | (((imm >> 20) & 0x1) << 31);
                bytecode = binary_imm | ((rd & 0x1F) << 7) | (opcode_number & 0x7F);
    
            } else if (strcmp(opcode, "JALR") == 0) {
                int rd = atoi(&words[1][1]);
                int rs1 = atoi(&words[2][1]);
                int imm = atoi(words[3]);
                imm = int2signedbin(imm, 12);
                int opcode_number = 0b1100111;
                bytecode = ((imm & 0xFFF) << 20) | ((rs1 & 0x1F) << 15) | ((get_funct3(opcode) & 0x7) << 12) | ((rd & 0x1F) << 7) | (opcode_number & 0x7F);
    
            } else if (strcmp(opcode, "BEQ") == 0 || strcmp(opcode, "BNE") == 0 ||
                       strcmp(opcode, "BLT") == 0 || strcmp(opcode, "BGE") == 0 ||
                       strcmp(opcode, "BLTU") == 0 || strcmp(opcode, "BGEU") == 0) {
                int rs2 = atoi(&words[1][1]);
                int rs1 = atoi(&words[2][1]);
                int imm = atoi(words[3]);
                imm = int2signedbin(imm, 30);
                int opcode_number = 0b1100011;
                int binary_imm1 = (((imm >> 11) & 0x1) << 7) | (((imm >> 1) & 0xF) << 8);
                int binary_imm2 = (((imm >> 5) & 0x3F) << 25) | (((imm >> 12) & 0x1) << 31);
                bytecode = binary_imm1 | binary_imm2 | ((rs2 & 0x1F) << 20) | ((rs1 & 0x1F) << 15) | ((get_funct3(opcode) & 0x7) << 12) | (opcode_number & 0x7F);
    
            } else if (strcmp(opcode, "LB") == 0 || strcmp(opcode, "LH") == 0 ||
                       strcmp(opcode, "LW") == 0 || strcmp(opcode, "LBU") == 0 ||
                       strcmp(opcode, "LHU") == 0) {
                int rd = atoi(&words[1][1]);
                int rs1 = atoi(&words[2][1]);
                int imm = atoi(words[3]);
                imm = int2signedbin(imm, 12);
                int opcode_number = 0b0000011;
                bytecode = ((imm & 0xFFF) << 20) | ((rs1 & 0x1F) << 15) | ((get_funct3(opcode) & 0x7) << 12) | ((rd & 0x1F) << 7) | (opcode_number & 0x7F);
    
            } else if (strcmp(opcode, "SB") == 0 || strcmp(opcode, "SH") == 0 ||
                       strcmp(opcode, "SW") == 0) {
                int rs2 = atoi(&words[1][1]);
                int rs1 = atoi(&words[2][1]);
                int imm = atoi(words[3]);
                imm = int2signedbin(imm, 12);
                int opcode_number = 0b0100011;
                bytecode = (((imm >> 5) & 0x7F) << 25) | ((rs2 & 0x1F) << 20) | ((rs1 & 0x1F) << 15) | ((get_funct3(opcode) & 0x7) << 12) | ((imm & 0x1F) << 7) | (opcode_number & 0x7F);
    
            } else if (strcmp(opcode, "ADDI") == 0 || strcmp(opcode, "SLTI") == 0 ||
                       strcmp(opcode, "SLTIU") == 0 || strcmp(opcode, "XORI") == 0 ||
                       strcmp(opcode, "ORI") == 0 || strcmp(opcode, "ANDI") == 0) {
                int rd = atoi(&words[1][1]);
                int rs1 = atoi(&words[2][1]);
                int imm = atoi(words[3]);
                imm = int2signedbin(imm, 12);
                int opcode_number = 0b0010011;
                bytecode = ((imm & 0xFFF) << 20) | ((rs1 & 0x1F) << 15) | ((get_funct3(opcode) & 0x7) << 12) | ((rd & 0x1F) << 7) | (opcode_number & 0x7F);
    
            } else if (strcmp(opcode, "SLLI") == 0 || strcmp(opcode, "SRLI") == 0 ||
                       strcmp(opcode, "SRAI") == 0) {
                int rd = atoi(&words[1][1]);
                int rs1 = atoi(&words[2][1]);
                int imm = atoi(words[3]);
                // Ensure imm >= for shifts
                assert(imm >= 0);
                int opcode_number = 0b0010011;
                bytecode = ((get_funct7(opcode) & 0x7F) << 25) | ((imm & 0x1F) << 7) | ((rs1 & 0x1F) << 12) | ((get_funct3(opcode) & 0x7) << 17) | ((rd & 0x1F) << 20) | ((opcode_number & 0x7F) << 25);
    
            } else if (strcmp(opcode, "ADD") == 0 || strcmp(opcode, "SUB") == 0 ||
                       strcmp(opcode, "SLL") == 0 || strcmp(opcode, "SLT") == 0 ||
                       strcmp(opcode, "SLTU") == 0 || strcmp(opcode, "XOR") == 0 ||
                       strcmp(opcode, "SRL") == 0 || strcmp(opcode, "SRA") == 0 ||
                       strcmp(opcode, "OR") == 0 || strcmp(opcode, "AND") == 0) {
                int rd = atoi(&words[1][1]);
                int rs1 = atoi(&words[2][1]);
                int rs2 = atoi(&words[3][1]);
                int opcode_number = 0b0110011;
                bytecode = ((get_funct7(opcode) & 0x7F) << 25) | ((rs2 & 0x1F) << 20) | ((rs1 & 0x1F) << 15) | ((get_funct3(opcode) & 0x7) << 12) | ((rd & 0x1F) << 7) | (opcode_number & 0x7F);
            }
        }
        
        // Write bytecode to output file
        char bits[9];

        byte2bits(bits, bytecode & 0xFF);
        fprintf(file_w, "Data_Rom(%d) <= \"%s\";\n", bytecounter++, bits);

        bytecode = bytecode >> 8;
        byte2bits(bits, bytecode & 0xFF);
        fprintf(file_w, "Data_Rom(%d) <= \"%s\";\n", bytecounter++, bits);

        bytecode = bytecode >> 8;
        byte2bits(bits, bytecode & 0xFF);
        fprintf(file_w, "Data_Rom(%d) <= \"%s\";\n", bytecounter++, bits);

        bytecode = bytecode >> 8;
        byte2bits(bits, bytecode & 0xFF);
        fprintf(file_w, "Data_Rom(%d) <= \"%s\";\n", bytecounter++, bits);
    }

    fclose(file_r);
    fclose(file_w);

    return 0;
}